common/autoLocker.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+)
+
+var PanicOnAutoLockFinalize = false
+
+// AutoLock provides a local object wrapper for lock management. It also provides a failsafe
+// finalizer that will unlock an autoLock when the object is garbage collected, which is similar
+// to, but less-performant than a C++ style auto-unlock destructor
+//
+// Use the finalizing version if you don't want to defer for performance reasons, but you want assurance
+// that your lock must be freed
+//
+// * B..b..but I need an RWMutex auto locker!
+//
+//   Pass your sync.RWMutex as-is for a W-autoLock, and use the sync.RWMutex.RLocker() for an R-autoLock
+// 
+// Example:
+// 
+// type foo struct {
+// 	l sync.RWMutex
+// }
+// 
+// func (f *foo) someFunc() {
+// 	l := AutoLock(&f.l.RLocker())
```

Comment:
the idea here is to have an object that can optionally be unlocked automatically if it leaves the scope, with the finalizer. What you are saying is reasonable if you don't need the finalizer as an insurance policy. Nothing precludes using this object in that way.